basic-services/ReadPropertyResult.cs

```diff
@@ -39,42 +40,130 @@ public struct ReadPropertyResult
         {
+            Id = id;
+            Attribute = attribute;
             Reliability = reliability ?? Reliable;
             Priority = priority;
-            
+            StringValue = null;
+            NumericValue = 1;
+            ArrayValue = null;
+            BooleanValue = false;
+            _CultureInfo = cultureInfo;
+
+            ReadToken(token);
+        }
+
+        /// <summary>Parses the JToken type and assigns struct values appropriately.</summary>
+        internal void ReadToken(JToken token)
+        {
+            if (token == null) {
+                NumericValue = 1;
+                StringValue = "Unsupported Data Type";
+                return;
+            }
             // switch on token type and set the fields appropriately
             switch (token.Type)
             {
                 case JTokenType.Integer:
                     NumericValue = token.Value<double>();
-                    StringValue = NumericValue.ToString(cultureInfo);
+                    StringValue = NumericValue.ToString(_CultureInfo);
                     BooleanValue = Convert.ToBoolean(NumericValue);
                     ArrayValue = null;
-
                     break;
                 case JTokenType.Float:
-                    //todo
+                    NumericValue = token.Value<double>();
+                    StringValue = NumericValue.ToString(_CultureInfo);
+                    BooleanValue = Convert.ToBoolean(NumericValue);
+                    break;
                 case JTokenType.String:
-                    // todo
+                    NumericValue = 0;
+                    StringValue = token.Value<string>();
+                    break;
                 case JTokenType.Array:
-                    // todo
+                    ReadArray(token);
+                    break;
                 case JTokenType.Boolean:
-                    // todo
+                    if ((bool)(token) == true) {
+                        NumericValue = 1;
+                        StringValue = "True";
+                        BooleanValue = true;
+                    } else {
+                        NumericValue = 0;
+                        StringValue = "False";
+                        BooleanValue = false;
+                    }
+                    break;
+                case JTokenType.Object:
+                    ReadObject(token);
+                    break;
                 default:
-                    StringValue = null;
-                    NumericValue = 0;
-                    ArrayValue = null;
-                    BooleanValue = false;
+                    NumericValue = 1;
+                    StringValue = "Unsupported Data Type";
                     break;
             }
-            
         }
 
-    }
-
-    
+        /// <summary>Parses a JArray and adds (string, double) pairs to an array based on JToken type.</summary>
```

Comment:
I think that the array should be an array of ReadPropertyResult values. That way

1. You have just as much support for elements as you do for one property. (like boolean support as well as string and double)
2. The interface for accessing values on an element is the same as the interface for an individual property
3. The code for instantiating the elements can be reused (ReadPropertyResult constructor)

Because of this proposed change we may want to rename ReadPropertyResult to something like `Variant` or `ValueVariant`. (*Variant* is a legacy term from old Visual Basic meaning that the value can be *anything*)

Reply:
I changed the array but did not rename the class yet.

Reply:
Will be renamed to Variant for now.